scipy/spatial/src/distance_pybind.cpp

```diff
@@ -0,0 +1,593 @@
+        const py::array& w_obj, WeightedDistanceFunc<scalar_t> f) {
+    auto x = npy_asarray<scalar_t>(x_obj,
+                                   NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
+    auto w = npy_asarray<scalar_t>(w_obj,
+                                   NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
+    auto out = py::cast<py::array_t<scalar_t>>(out_obj);
+    auto out_desc = get_descriptor(out);
+    auto out_data = out.mutable_data();
+    auto x_desc = get_descriptor(x);
+    auto x_data = x.data();
+    auto w_desc = get_descriptor(w);
+    auto w_data = w.data();
+    {
+        py::gil_scoped_release guard;
+        validate_weights(w_desc, w_data);
+        pdist_weighted_impl(
+            out_desc, out_data, x_desc, x_data, w_desc, w_data, f);
+    }
+    return out;
+}
+
+template <typename scalar_t>
+py::array cdist_unweighted(const py::array& out_obj, const py::array& x_obj,
+                           const py::array& y_obj, DistanceFunc<scalar_t> f) {
+    auto x = npy_asarray<scalar_t>(x_obj,
+                                 NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
+    auto y = npy_asarray<scalar_t>(y_obj,
+                                 NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
+    auto out = py::cast<py::array_t<scalar_t>>(out_obj);
+
+    auto out_desc = get_descriptor(out);
+    auto out_data = out.mutable_data();
+    auto x_desc = get_descriptor(x);
+    auto x_data = x.data();
+    auto y_desc = get_descriptor(y);
+    auto y_data = y.data();
+    {
+        py::gil_scoped_release guard;
+        cdist_impl(out_desc, out_data, x_desc, x_data, y_desc, y_data, f);
+    }
+    return out;
+}
+
+template <typename scalar_t>
+py::array cdist_weighted(
+        const py::array& out_obj, const py::array& x_obj,
+        const py::array& y_obj, const py::array& w_obj,
+        WeightedDistanceFunc<scalar_t> f) {
+    auto x = npy_asarray<scalar_t>(x_obj,
+                                 NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
+    auto y = npy_asarray<scalar_t>(y_obj,
+                                 NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
+    auto w = npy_asarray<scalar_t>(w_obj,
+                                 NPY_ARRAY_ALIGNED | NPY_ARRAY_NOTSWAPPED);
+    auto out = py::cast<py::array_t<scalar_t>>(out_obj);
+
+    auto out_desc = get_descriptor(out);
+    auto out_data = out.mutable_data();
+    auto x_desc = get_descriptor(x);
+    auto x_data = x.data();
+    auto y_desc = get_descriptor(y);
+    auto y_data = y.data();
+    auto w_desc = get_descriptor(w);
+    auto w_data = w.data();
+    {
+        py::gil_scoped_release guard;
+        validate_weights(w_desc, w_data);
+        cdist_weighted_impl(
+            out_desc, out_data, x_desc, x_data, y_desc, y_data, w_desc, w_data, f);
+    }
+    return out;
+}
+
+py::dtype npy_promote_types(const py::dtype& type1, const py::dtype& type2) {
+    // TODO: Casts away const, but think NumPy API should allow const here?
```

Comment:
My guess for the lack of const is that there may be internal bits of a dtype that gets mutated by numpy without changing what it actually stands for, e.g. lazy-attributes initialized on first access. If that's the case then the const above may need to be removed.

Reply:
Looked into it, it uses `Py_INCREF` internally mutates the object's reference count. However, `.ptr()` actually returns a non-const pointer anyway so this doesn't cast away const at all.